src/Renderer/LayeredMaterial.ts

```diff
@@ -0,0 +1,542 @@
+    max: number,
+) {
+    // Aliases for readability
+    const uLayers = uniforms.layers.value;
+    const uTextures = uniforms.textures.value;
+    const uOffsetScales = uniforms.offsetScales.value;
+    const uTextureCount = uniforms.textureCount;
+
+    // Flatten the 2d array: [i, j] -> layers[_layerIds[i]].textures[j]
+    let count = 0;
+    for (const tile of tiles) {
+        // FIXME: RasterElevationTile are always passed to this function alone
+        // so this works, but it's really not great even ignoring the dynamic
+        // addition of a field.
+        // @ts-expect-error: adding field to passed layer
+        tile.textureOffset = count;
+
+        for (
+            let i = 0;
+            i < tile.textures.length && count < max;
+            ++i, ++count
+        ) {
+            uOffsetScales[count] = tile.offsetScales[i];
+            uTextures[count] = tile.textures[i];
+            uLayers[count] = tile;
+        }
+    }
+    if (count > max) {
+        console.warn(
+            `LayeredMaterial: Not enough texture units (${max} < ${count}),`
+            + 'excess textures have been discarded.',
+        );
+    }
+    uTextureCount.value = count;
+}
+
+export const ELEVATION_MODES = {
+    RGBA: 0,
+    COLOR: 1,
+    DATA: 2,
+};
+
+/**
+ * Convenience type that wraps all of the generic type's fields in
+ * [THREE.IUniform]s.
+ */
+type MappedUniforms<Uniforms> = {
+    [name in keyof Uniforms]: THREE.IUniform<Uniforms[name]>;
+};
+
+/** List of the uniform types required for a LayeredMaterial. */
+interface LayeredMaterialRawUniforms {
+    // Color
+    diffuse: THREE.Color;
+    opacity: number;
+
+    // Lighting
+    lightingEnabled: boolean;
+    lightPosition: THREE.Vector3;
+
+    // Misc
+    fogDistance: number;
+    fogColor: THREE.Color;
+    overlayAlpha: number;
+    overlayColor: THREE.Color;
+    objectId: number;
+    geoidHeight: number;
+
+    // > 0 produces gaps,
+    // < 0 causes oversampling of textures
+    // = 0 causes sampling artefacts due to bad estimation of texture-uv
+    // gradients
+    // best is a small negative number
+    minBorderDistance: number;
+
+    // Debug
+    showOutline: boolean,
+    outlineWidth: number,
+    outlineColors: THREE.Color[]
+
+    // Elevation layers
+    elevationLayers: Array<StructElevationLayer>,
+    elevationTextures: Array<THREE.Texture>,
+    elevationOffsetScales: Array<THREE.Vector4>,
+    elevationTextureCount: number,
+
+    // Color layers
+    colorLayers: Array<StructColorLayer>,
+    colorTextures: Array<THREE.Texture>,
+    colorOffsetScales: Array<THREE.Vector4>,
+    colorTextureCount: number,
+}
+
+let nbSamplers: [number, number] | undefined;
+const fragmentShader: string[] = [];
+
+/** Replacing the default uniforms dynamic type with our own static map. */
+export type LayeredMaterialParameters =
+    Omit<THREE.ShaderMaterialParameters, 'uniforms'>
+    & { uniforms?: MappedUniforms<LayeredMaterialRawUniforms> };
```

Comment:
I think it would be easier to read as an extended interface:

```suggestion
export interface LayeredMaterialParameters extends THREE.ShaderMaterialParameters {
    uniforms?: MappedUniforms<LayeredMaterialRawUniforms>;
}
```

Also, do we authorize user of the `LayeredMaterial` class to set its own defines/uniforms?

Reply:
I don't see a reason to allow custom defines/uniforms given that the pipeline used is fixed (unless the user goes and modifies the shaders, which would be as unintended as can be).

Reply:
In that case, you should maybe omit both `uniforms` and `defines` properties.